hbase-mapreduce/src/main/java/org/apache/hadoop/hbase/mob/mapreduce/MobRefReporter.java

```diff
@@ -0,0 +1,485 @@
+              context.write(HLINK_RESTORE, key);
+            } else {
+              LOG.warn("Found file '{}' in archive area, but the hlink back references do not "
+                  + "properly point to the mob area for our table.", file);
+              context.write(ARCHIVE_ERROR_BAD_LINK, encodeRows(context, key, rows));
+            }
+          } else {
+            LOG.warn("Found file '{}' in archive area, but there are no hlinks pointing to it. Not "
+                + "yet used snapshot or an error.", file);
+            context.write(ARCHIVE_ERROR_STALE, encodeRows(context, key, rows));
+          }
+        } catch (IOException e) {
+          LOG.warn("Found file '{}' in archive area, but got an error while checking "
+              + "on back references.", file, e);
+          context.write(ARCHIVE_ERROR_IO, encodeRows(context, key, rows));
+        }
+
+      } else {
+        // hlink in active mob area (from a snapshot of a different table)
+        // we're doing this ourselves instead of using FSUtils.getReferenceFilePaths because
+        // we know the mob region never splits, so we can only have HFileLink references
+        // and looking for just them is cheaper then listing everything.
+        try {
+          FileStatus[] hlinks = mob.getFileSystem(conf).globStatus(new Path(mob + "/*=*-" + file));
+          if (hlinks != null && hlinks.length != 0) {
+            if (hlinks.length != 1) {
+              LOG.warn("Found file '{}' as hfilelinks in the mob area, but there are more than " +
+                  "one: {}", file, Arrays.deepToString(hlinks));
+            }
+            HFileLink found = null;
+            for (FileStatus hlink : hlinks) {
+              HFileLink tmp = HFileLink.buildFromHFileLinkPattern(conf, hlink.getPath());
+              if (tmp.exists(archive.getFileSystem(conf))) {
+                found = tmp;
+                break;
+              } else {
+                LOG.debug("Target file doesn't exist for ref {}", tmp);
+              }
+            }
+            if (found != null) {
+              LOG.debug("Found file '{}' as a ref in the mob area: {}", file, found);
+              context.write(HLINK_CLONE, key);
+            } else {
+              LOG.warn("Found file '{}' as ref(s) in the mob area but they don't point to an hfile "
+                  + "that exists.", file);
+              context.write(HLINK_ERROR_DANGLING, encodeRows(context, key, rows));
+            }
+          } else {
+            LOG.error("Could not find referenced file '{}'.", file);
+            LOG.debug("Note that we don't have the server-side tag from the mob cells that says "
+                + "what table the reference is originally from. So if the HFileLink in this table "
+                + "is missing but the referenced file is still in the table from that tag, then "
+                + "lookups of these impacted rows would still work.");
```

Comment:
Is this a case an administrator would need to take action on or is it possibly really bad and possibly perfectly benign?

Reply:
definitely need to take action. it's either dataloss or dataloss waiting to happen. Let me include how to differentiate.

Reply:
how's the added text to the MISSING FILE description on the tool output?

Reply:
Added a comment up by `MISSING FILE`. Should it be `would still work` to `will work`; maybe suggest the user do a GET on that cell to verify (or can the tool do that verification in an "online" mode)?

Reply:
I could do a get. Lemme see about that. Not sure if we still have info for an hbase connection once we're in the reducer.

Reply:
Okay I chatted with @esteban about this a bit.

One issue is that we wouldn't be able to do a simple GET because we only have a row key and a family. Doing a scan is also complicated because a given row could have multiple MOB cells that point to different files and those files could be a mix of missing and present-but-improperly accounted for. I think that means if a given impacted row gave an error when scanning, we'd still have to scan the rest of the rows. and even after getting an error on every impacted row we wouldn't know if the errors were because of a cell that pointed at this particular file or if it was because of another cell in the family that pointed at a different file.

What if instead we could have the driver build a listing of all `hfile -> table` so that we can derive the information we're missing by not having the server side tag? We do this once and then put the result in the distributed cache. When a reducer needs to confirm that the file is gone-gone we can check the lookup. For that to have a race condition we'd need to somehow start the job then replace the table via a snapshot we took of a different table that flushed data after we built the listing.

If I grab a directory listing from an example cluster it looks like this lookup table is ~11m entries, taking up about 196MB gzipped and 1.1GB uncompressed.